src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java

```diff
@@ -235,37 +245,61 @@ public void authenticateAndAuthorizeGitRequest(HttpServletRequest request, Repos
         log.debug("Authorizing user {} for repository {} took {}", user.getLogin(), localVCRepositoryUri, TimeLogUtil.formatDurationFrom(timeNanoStart));
     }
 
-    private User authenticateUser(String authorizationHeader) throws LocalVCAuthException {
+    private User authenticateUser(String authorizationHeader, ProgrammingExercise exercise, LocalVCRepositoryUri localVCRepositoryUri) throws LocalVCAuthException {
 
         UsernameAndPassword usernameAndPassword = extractUsernameAndPassword(authorizationHeader);
 
         String username = usernameAndPassword.username();
         String password = usernameAndPassword.password();
-
-        var user = userRepository.findOneByLogin(username);
+        User user = userRepository.findOneByLogin(username).orElseThrow(LocalVCAuthException::new);
 
         try {
             SecurityUtils.checkUsernameAndPasswordValidity(username, password);
+        }
+        catch (AccessForbiddenException | AuthenticationException e) {
+            throw new LocalVCAuthException();
+        }
 
-            // Note: we first check if the user has used a vcs access token instead of a password
+        // check user VCS access token
+        if (Objects.equals(user.getVcsAccessToken(), password) && user.getVcsAccessTokenExpiryDate() != null && user.getVcsAccessTokenExpiryDate().isAfter(ZonedDateTime.now())) {
```

Comment:
Question (just to be sure): SecurityUtils.checkUsernameAndPasswordValidity checks that the password is not empty, right? Otherwise, this call would be insecure (i.e., password == null would be authenticated)

Reply:
As far as I can see that is checked here: https://github.com/ls1intum/Artemis/blob/c29ca89224027c50b5294665f6377f27156a0e29/src/main/java/de/tum/in/www1/artemis/security/SecurityUtils.java#L41 so should be fine

Reply:
yep, empty strings are not allowed